packages/microcosm-www-next/src/pages/index.js

```diff
@@ -1,40 +1,152 @@
+    this.graphicsMap = this.setGraphicsMap(graphics)
+  }
+
+  setGraphicsMap(graphics) {
+    return [].slice.call(graphics).reduce((map, graphic) => {
+      map[graphic.dataset.section] = graphic
+      return map
+    }, {})
+  }
+
+  beginObserve() {
+    //create new Observer instance
+    let observer = new IntersectionObserver(
+      this.onIntersection,
+      this.observeOptions
+    )
+
+    //start observing each graphic
+    Object.values(this.graphicsMap).forEach(graphic =>
+      observer.observe(graphic)
+    )
+  }
+
+  onIntersection = observed => {
+    let entry = observed[0]
+    let section = entry.target.dataset.section
+    let notAlreadyVisible = section !== this.currentSection
+
+    if (entry.isIntersecting && notAlreadyVisible) {
+      this.currentSection = section
+      this.changeHeading()
+      this.changeText()
+      this.changeGraphic()
+    }
+  }
+
+  changeHeading() {
+    let sectionData = data[this.currentSection]
+    this.heading.innerHTML = sectionData.heading
+  }
+
+  changeText() {
+    let sectionData = data[this.currentSection]
+    this.text.innerHTML = this.onMicrocosmView
+      ? sectionData.microcosmText
+      : sectionData.browserText
+  }
+
+  changeGraphic() {
+    let section = this.currentSection
+    let graphicElement = this.graphicsMap[section]
+    let newGraphicUrl = this.onMicrocosmView
+      ? data[section].microcosmGraphicUrl
+      : data[section].browserGraphicUrl
+
+    graphicElement.setAttribute('src', newGraphicUrl)
+  }
+
+  changeButtonText(button) {
+    button.classList.toggle('-browserView', !this.onMicrocosmView)
+  }
+
+  changeSubheadingText(button) {
+    this.subheadingTop.classList.toggle('-browserView', !this.onMicrocosmView)
+    this.subheadingBottom.classList.toggle('-browserView', !this.onMicrocosmView)
+  }
+
+  switchView = e => {
+    this.onMicrocosmView = !this.onMicrocosmView
+
+    this.changeGraphic()
+    this.changeSubheadingText()
+    this.changeText()
+    this.changeButtonText(e.target)
+  }
+
+  render() {
+    return (
+      <div className="wrapper">
+        <section className="section">
+          <div className="section__content">
+            <h2 className="section__content__heading" id="heading">
+              <span>01.</span>
+              Rendering a View
+            </h2>
+
+            <h3
+              className="section__content__subheading -top"
+              id="subheading-top"
+            >
+              In
+            </h3>
+            <p className="section__content__text" id="text">
               The <a href="TODO">Domains</a> are in charge of keeping state
               organized, and provide whatever data is necessary to the
               Presenter. A Presenter at it's core is a React Component, so it
               uses the data it pulls from the Domains to render the appropriate
               view.
             </p>
-          </figcaption>
-        </figure>
-      </div>
-    </div>
-  </section>
-)
```

Comment:
There is a lot of direct DOM manipulation in this component. If this component renders again, React might reverse changes you've made manually.

Instead of assigning state directly to this class, what if you used `setState`? Then you could lean on React to do the DOM manipulation for you.

So instead of something like `this.changeHeading()`, you'd just do `this.setState({ heading: "..." })`. That would cause React to do a re-render for you, updating the heading on your behalf.

Reply:
The other benefit is that then all of your state for tracking what heading or graphic is displaying lives inside of the React component. You don't need to read from the DOM to get the current state of things. For me, this makes it much simpler to reason about rendering a page.

The `render` method declares the ideal state. Once you have that set up, all you need to do is change a components state via `this.setState`. Then React calls render again, and does all of the DOM manipulation to get your page to show the next set of content.

Reply:
yes, very good call!! I totally forgot about `setState` so I will take advantage of that